### DISCUSSION TITLE
Issues scaling servers: network is already full

### DISCUSSION THREAD

**Discussion by Niek** [mysticaltech/terraform-hcloud-kube-hetzner#730] [Q&A]
I have an autoscaler group that I can't seem to grow above 94 instances. The error I see in the `cluster-autoscaler-x` logs is:
```
E0417 10:47:10.721880       1 hetzner_node_group.go:117] failed to create error: could not create server type cpx21 in region ash: network 1234567 is already full (invalid_input)
```

If I check in the Hetzner UI I see only 95 resources/routes (94 + control plane), that does't seem too much to me? Any idea what to do?

**Comment by ifeulner** [comment]
Contact the Hetzner support, maybe they are reserving additional IPs for the node internally... Seems your cluster is quite big.

Let us know the outcome, please!

**Reply by Niek** [reply]
I got this reply:
> There are a maximum of 100 Cloud-Servers per network.
>
> You can find the limits for the private networks here:
> https://docs.hetzner.com/de/cloud/networks/overview#limits
>
> Please note that this is a hard-limit, which cannot be adjusted. If you wish to connect more than 100 Servers, you would need to create an additional Network.

So, in order to scale to 100+ nodes, it seems that multiple networks must be used.

**Reply by mysticaltech** [reply]
Oh my! That's a significant issue then, meaning our project is limited to 100 nodes.

@kube-hetzner/core FYI, if you have ideas. It seems like a hard problem because, unless we create multiple networks and interconnect them with a CNI with wireguard. However, we will have to manage the routes between networks ourselves, it seems like a mess.

**Reply by aleksasiriski** [reply]
The real question is: Why is that a hard limit on Hetzner side?

Also @Niek as a workaround, maybe you can divide your workflow into multiple clusters? And dang, if I may ask what is your use case for such a huge cluster!?

**Reply by Niek** [reply]
Yeah, for now multiple clusters seems like the easiest solution. I run rather small `cpx21` nodes, I could also go with much larger nodes (like `cpx51`) and run more pods per node on it.
I think this section requires some updating though, as it's much more limited than thought: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner#scaling-nodes

**Reply by mysticaltech** [reply]
@Niek Definitely vertical scaling is the solution here, you can have much much bigger nodes, especially now that we support ARM instances, you can save a lot of 💵.

**Reply by mysticaltech** [reply]
I will update the docs ASAP too.

**Reply by ifeulner** [reply]
That's really a pity - I always thought this is per subnet, but obviously it's per network. And that's quite a limitation, at least compared to the big hyperscalers.

**Reply by ifeulner** [reply]
> Yeah, for now multiple clusters seems like the easiest solution. I run rather small `cpx21` nodes, I could also go with much larger nodes (like `cpx51`) and run more pods per node on it.
> I think this section requires some updating though, as it's much more limited than thought: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner#scaling-nodes

In general this is not a good practice, as it raises extremely the internode-communication. So it's better to find a sweet spot between node size and number of nodes depending on your requirements. It could also make sense to increase the number of pods per node (200 by default I think...) .

**Reply by mysticaltech** [reply]
Good point @ifeulner, however 163-32GB nodes seems obviously ok to use. @Niek If I were you, I would shoot for the CAX41 instances, 16 vCPU / 32GB @ 29 euros!

**Comment by mysticaltech** [comment]
@apricot FYI above, some of our users need more than 100 nodes and it seems that private networks do not support that. It seems like a really low limit. Any recommendations on getting around that? And are you guys planning on scaling that number up?
